code/scripting/scripting.cpp

```diff
@@ -345,249 +346,284 @@ bool ConditionedHook::ConditionsValid(int action, object *objp, int more_data)
+				{
+					if (objp != nullptr && objp->type == OBJ_SHIP)
+					{
+						auto sip = &Ship_info[Ships[objp->instance].ship_info_index];
+						if (sip->class_type >= 0)
+						{
+							if (stricmp(Ship_types[sip->class_type].name, scp.condition_string.c_str()) != 0)
+								return false;
+						}
+						break;
+					}
+				}
 				break;
+			}
+
 			case CHC_SHIPCLASS:
-				//scp.condition_cached_value holds the weapon_info_index of the requested weapon class
-				if(objp == NULL || objp->type != OBJ_SHIP)
+			{
+				for (auto objp : objp_array)
+				{
+					if (objp != nullptr && objp->type == OBJ_SHIP)
+					{
+						// scp.condition_cached_value holds the ship_info_index of the requested ship class
+						if (Ships[objp->instance].ship_info_index != scp.condition_cached_value)
+							return false;
+						break;
+					}
+				}
+				break;
+			}
+
+			case CHC_SHIP:
+			{
+				for (auto objp : objp_array)
+				{
+					if (objp != nullptr && objp->type == OBJ_SHIP)
+					{
+						if (stricmp(Ships[objp->instance].ship_name, scp.condition_string.c_str()) != 0)
+							return false;
+						break;
+					}
+				}
+				break;
+			}
+
+			case CHC_MISSION:
+			{
+				//WMC - Get mission filename with Mission_filename
+				//I don't use Game_current_mission_filename, because
+				//Mission_filename is valid in both fs2_open and FRED
+				size_t len = strlen(Mission_filename);
+				if(!len)
 					return false;
-				if(!(Ships[objp->instance].ship_info_index == scp.condition_cached_value))
+				if(len > 4 && !stricmp(&Mission_filename[len-4], ".fs2"))
+					len -= 4;
+				if(strnicmp(scp.condition_string.c_str(), Mission_filename, len) != 0)
 					return false;
 				break;
-			case CHC_SHIP:
-				if(objp == NULL || objp->type != OBJ_SHIP)
+			}
+
+			case CHC_CAMPAIGN:
+			{
+				size_t len = strlen(Campaign.filename);
+				if(!len)
 					return false;
-				if(stricmp(Ships[objp->instance].ship_name, scp.condition_string.c_str()) != 0)
+				if(len > 4 && !stricmp(&Mission_filename[len-4], ".fc2"))
+					len -= 4;
+				if(strnicmp(scp.condition_string.c_str(), Mission_filename, len) != 0)
 					return false;
 				break;
-			case CHC_MISSION:
+			}
+
+			case CHC_WEAPONCLASS:
+			{
+				for (auto objp : objp_array)
 				{
-					//WMC - Get mission filename with Mission_filename
-					//I don't use Game_current_mission_filename, because
-					//Mission_filename is valid in both fs2_open and FRED
-					size_t len = strlen(Mission_filename);
-					if(!len)
-						return false;
-					if(len > 4 && !stricmp(&Mission_filename[len-4], ".fs2"))
-						len -= 4;
-					if(strnicmp(scp.condition_string.c_str(), Mission_filename, len) != 0)
-						return false;
-					break;
+					if (objp != nullptr && (objp->type == OBJ_WEAPON || objp->type == OBJ_BEAM))
+					{
+						// scp.condition_cached_value holds the weapon_info_index of the requested weapon class
+						if (Weapons[objp->instance].weapon_info_index != scp.condition_cached_value)
```

Comment:
There will need to be a separate check for `OBJ_BEAM` in the `Beams` array since I don't believe those have a weapon object at all (making this an invalid access).

Reply:
Good catch.  Updated.